### PR TITLE
First pass at the requirements

### PR DESCRIPTION
**Pull request by kariabancroft**
## Description
This project will replace farmar in our week of curriculum where we have three days to wrap up all of the Ruby work we've done so far.

## TODO
- [x] Create CSV files
- [x] Write optional requirements
- [x] Add code coverage requirements
- [x] Add feedback template
- [x] Confirm comprehension questions after requirements are finalized

## Feedback Requested
**Scope**: Do we feel like this will be doable in three days (plus a few weekend hours)?
**Leveling**: Ideally this is more open-ended than farmar, requiring more design work and more flexibility and options
**Learning Goals**: I tried to hone in on our actual learning goals rather than some of the things that I felt were more arbitrary difficult in farmar (file structure, for example)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,84 @@
...
+- Writing pseudocode and creating tests to drive the creation of our code
+- Importing data from a CSV file
+
+## Primary Requirements
+Create a system that will manage information related to ride sharing. The system should have a way to keep track of drivers, riders and the trips that the drivers take with their riders.
+
+
+#### Driver
+- Each driver should have information uniquely identifying them like ID, name, license plate and vehicle identification number
+- Each vehicle identification number should abide by a specific pattern or length to ensure it is a valid vehicle identification number
```
Comment (Hamled) [inline README.md]: The provided dataset has VINs that are clearly following a specific pattern already. Should we tell the students what pattern to validate against (while requiring them to figure out how to code it)?

While I don't think reverse engineering the pattern from the given dataset would be a struggle for any of our students, it's not a clear part of the learning goals for this project and they're already short on time.

File: README.md [hunk 2]
```diff
@@ -1 +1,84 @@
...
+Remember the ride share exercise we did with designing and creating a system to track the ride share data from a CSV file? We did a lot of great work on this exercise in creating arrays and hashes of data, but we've learned a lot since we did that exercise! Now, we're going to use our understanding of classes, methods and attributes to create an object-oriented implementation of our ride share system.
+
+## Learning Goals
+Reinforce and practice all of the Ruby and programming concepts we've covered in class so far:
+- Creating and instantiating classes with attributes
+- Creating class and instance methods within our classes
+- Writing pseudocode and creating tests to drive the creation of our code
+- Importing data from a CSV file
+
+## Primary Requirements
```
Comment (Hamled) [inline README.md]: I think the requirements section could be structured a bit more, by having multiple levels in the requirement lists.

Where we're saying "You should be able to" multiple times in a row we can use that prefix once and nest the various requirements under it.

File: README.md [hunk 3]
```diff
@@ -1 +1,84 @@
...
+
+After a few minutes, then you'll get together with your seat squad to discuss the nouns and verbs you came up with.
+
+#### Phase 2
+Use your list of nouns to begin thinking about what classes you will want to use. Use your list of verbs associated with nouns to consider what methods you need to put in which classes.
+
+After a few minutes, then you'll get together with your seat squad to discuss the classes and methods you came up with.
+
+#### Phase 3
+Create a diagram that describes how each of these classes and methods (messages) will interact with one another as well as the CSV files.
```
Comment (Hamled) [inline README.md]: We use the term "message" only briefly in a couple of places in the rest of our curriculum, is it important to reinforce that usage here?

File: README.md [hunk 4]
```diff
@@ -1 +1,84 @@
...
+This is a level 3, individual project.
+
+We will use the same project structure we used for the previous project. Classes should be in files in the `lib` folder, and tests should be in files in the `specs` folder. You should utilize a spec helper file. You will run tests by executing the `rake` command, as configured in a Rakefile. You should have **95% code coverage** using simplecov.
+
+The `support` folder contains CSV files which will drive your system design. Each CSV corresponds to a different type of object _as well as_ creating a relationship between different objects.
+
+## Baseline
+### Discussion
+#### Phase 1
+Let's start by **individually** creating the list of nouns and verbs that you would utilize to describe this problem.
```
Comment (Hamled) [inline README.md]: I don't think there's any way around it really, but we do effectively give them all of the nouns in the requirements above.

File: README.md [hunk 5]
```diff
@@ -1 +1,84 @@
...
+Use your list of nouns to begin thinking about what classes you will want to use. Use your list of verbs associated with nouns to consider what methods you need to put in which classes.
+
+After a few minutes, then you'll get together with your seat squad to discuss the classes and methods you came up with.
+
+#### Phase 3
+Create a diagram that describes how each of these classes and methods (messages) will interact with one another as well as the CSV files.
+
+### Setup
+1. You will fork this repository in GitHub
+1. You will git clone the repository to your computer
```
Comment (Hamled) [inline README.md]: For these first two steps we should just say "Fork this.." and "Clone the..." to match the style of the following steps.

File: README.md [hunk 6]
```diff
@@ -1 +1,84 @@
...
+You should use the following process as much as possible:  
+
+1. Write pseudocode
+1. Write test(s)
+1. Write code
+
+Since this is the **last** Ruby-only project we are doing, and a level 3, the requirements are larger than previous projects. It is possible you will not be able to complete all requirements, but if you use the process above, you will still be able to share with your instructors the process and the design you have created.
+
+### Optional Requirements
+- The first optional should always be looking for improvements within the existing code you have already written. Are there any additional edge cases you might be missing in your tests? Any opportunity to use an enumerable method where you're using an each?
```
Comment (Hamled) [inline README.md]: If this should always be the first optional, are we planning to update future project READMEs to match?

File: README.md [hunk 7]
```diff
@@ -1 +1,84 @@
...
+- Enhance the trip functionality to include support for cost and duration
+- For a given rider, add the ability to return the total amount of money they have spent on all trips
+- For a given rider, add the ability to return the total amount of time they have spent on their trips
+- For a given driver, calculate their total revenue for all trips. Each driver gets 80% of the trip cost _after_ a fee of $1.65 is subtracted.
```
Comment (Hamled) [inline README.md]: I appreciate that we're more equitable to our drivers than Uber. 😆

File: README.md [hunk 8]
```diff
@@ -1 +1,84 @@
...
+## Learning Goals
+Reinforce and practice all of the Ruby and programming concepts we've covered in class so far:
+- Creating and instantiating classes with attributes
+- Creating class and instance methods within our classes
+- Writing pseudocode and creating tests to drive the creation of our code
+- Importing data from a CSV file
+
+## Primary Requirements
+Create a system that will manage information related to ride sharing. The system should have a way to keep track of drivers, riders and the trips that the drivers take with their riders.
+
```
Comment (Hamled) [inline README.md]: Should we add a note that a command-line interface is explicitly not part of the requirements?

We haven't really emphasized CLIs at all since Bank Accounts so maybe they wouldn't think to write that in?

**Comment by Hamled** [comment]
I don't see any obvious way around this, especially given the time constraints for project time next week, but are we at all concerned that the design work involved in this will feel relatively rote because they have such a clear example of a similar design in Bank Accounts?

File: .github/PULL_REQUEST_TEMPLATE [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+Congratulations! You're submitting your assignment!
+
+## Comprehension Questions
+| Question | Answer    |
+| :------------- | :------------- |
+| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |        |
+| Describe a concept that you gained more clarity on as you worked on this assignment. |    |
+| Describe a _nominal test_ that you wrote for this assignment. |    |
+| Describe an _edge case test_ that you wrote for this assignment. |    |
+| How do you feel you did in writing pseudocode first, then writing the tests and then the code? |   |
```
Comment (kariabancroft) [inline .github/PULL_REQUEST_TEMPLATE]: Not individually, the process as a whole

File: README.md (hunk 8 above)
Comment (kariabancroft) [inline README.md]: I don't think we need to say that explicitly b/c we didn't have an issue with that at all in Scrabble

File: README.md (hunk 1 above)
Comment (kariabancroft) [inline README.md]: That's why i said "pattern or length". At a minimum, they validate that it is at least the right # of characters. I could make that more clear as the MVP if you think that would be helpful.

File: README.md (hunk 4 above)
Comment (kariabancroft) [inline README.md]: Agreed - I think that's the only option

File: README.md (hunk 3 above)
Comment (kariabancroft) [inline README.md]: It's in the Metz reading which is why I want to continue using it here

File: README.md (hunk 6 above)
Comment (kariabancroft) [inline README.md]: sure?

**Comment by kariabancroft** [comment]
@Hamled I actually disagree that the design is that similar to bank accounts. There is still significant class "joining" that needs to be done that we never did at all with bank accounts.

**Comment by kariabancroft** [comment]
@Hamled Updated per your comments

File: README.md (hunk 1 above)
Comment (Hamled) [inline README.md]: I think it would be helpful for the students to see that the requirement was for a VIN to be exactly X characters. Then maybe we could add an optional for validating it more strictly.

**Comment by Hamled** [comment]
@kariabancroft Ah okay, I was thinking of the `Owner` class optional requirement from Wave 2 of Bank Accounts. But in my head I was sort of automatically extending that to have the account ownership relationship be its own class, which it definitely wasn't in that project.

File: README.md (hunk 1 above)
Comment (kariabancroft) [inline README.md]: Ok - updated the requirement. they can figure out what that length is.
